packages/ses/src/compartment-shim.js

```diff
@@ -1,33 +1,115 @@
+/* eslint no-underscore-dangle: ["off"] */
 
 // This module exports both Compartment and StaticModuleRecord because they
 // communicate through the moduleAnalyses private side-table.
 import {
   assign,
   create,
   defineProperties,
+  entries,
   freeze,
   getOwnPropertyNames,
   getOwnPropertyDescriptors,
+  keys,
 } from './commons.js';
 import { initGlobalObject } from './global-object.js';
 import { performEval } from './evaluate.js';
 import { isValidIdentifierName } from './scope-constants.js';
 import { sharedGlobalPropertyNames } from './whitelist.js';
-import { InertCompartment } from './inert.js';
+import { InertCompartment, InertStaticModuleRecord } from './inert.js';
 import {
   evadeHtmlCommentTest,
   evadeImportExpressionTest,
   rejectSomeDirectEvalExpressions,
 } from './transforms.js';
+import { load } from './module-load.js';
+import { link } from './module-link.js';
+import { getDeferredExports } from './module-proxy.js';
+import { assert } from './error/assert.js';
+
+const { quote: q } = assert;
+
+// moduleAliases associates every public module exports namespace with its
+// corresponding compartment and specifier so they can be used to link modules
+// across compartments.
+// The mechanism to thread an alias is to use the compartment.module function
+// to obtain the exports namespace of a foreign module and pass it into another
+// compartment's moduleMap constructor option.
+const moduleAliases = new WeakMap();
 
 // privateFields captures the private state for each compartment.
 const privateFields = new WeakMap();
 
+export const StaticModuleRecordPrototype = {
+  constructor: InertStaticModuleRecord,
+  toString() {
+    return '[object StaticModuleRecord]';
+  },
+};
+
+// eslint-disable-next-line no-shadow
+export const __PrecompiledStaticModuleRecord__ = function StaticModuleRecord(
+  analysis,
+) {
+  if (new.target === undefined) {
+    return new __PrecompiledStaticModuleRecord__(analysis);
+  }
+
+  const {
+    imports,
+    functorSource,
+    fixedExportMap,
+    liveExportMap,
+    exportAlls,
+  } = analysis;
+
+  // `keys` below is Object.keys which shows only the names of string-named
+  // enumerable own properties.
+  // By contrast, Reflect.ownKeys also shows the names of symbol-named
+  // enumerable own properties.
+  // `sort` defaults to a comparator that stringifies the array elements in a
+  // manner which fails on symbol-named properties.
+  // Distinct symbols can have the same stringification.
+  //
+  // The other subtle reason this is correct is that analysis.imports should
+  // only have identifier-named own properties.
+  this.imports = freeze(keys(imports).sort());
+  this.__functorSource__ = functorSource;
+  // We do not evidently need the values of this record, only the keys.
+  // this.__imports__ = freeze(imports);
+  this.__fixedExportMap__ = freeze(fixedExportMap);
+  this.__liveExportMap__ = freeze(liveExportMap);
+  this.__exportAlls__ = freeze(exportAlls);
```

Comment:
I'm also concerned about the security implications of exposing all these internals. Why is this not dangerous? (I think I see why. But given the nature of the hazard, it is worth an extended comment making the argument explicitly.)

Reply:
The expense is coupling. Using a WeakMap couples the StaticModuleRecord and Compartment constructors, such that they have to be in the same package if not the same module. The WeakMap could be made into public API to break the coupling, which would defeat the purpose of using a WeakMap. So, I’m going in the direction of a duck-type and making the duck-type as obviously non-standards-track as possible.

The metadata is information to the linker. A lie will result in a link failure. The functor source gets evaluated in a compartment with the power to update its dependees and receive updates from dependencies. I think that’s the extent of the security story.

Reply:
Oh, and the value of decoupling these is that the StaticModuleRecord is the only part that entrains Babel. In a subsequent change, it can be moved into another package, namely `@agoric/transform-module`, which I propose we rename to `@endo/static-module-record` to make the connection clear.

Reply:
> A lie will result in a link failure. ... I think that’s the extent of the security story.

I wasn't so much concerned with the static module record itself lying, though I probably should have worried about that too.

Rather, can adversarial code, with access to a static module record with these exposed internals, do any damage it could not do otherwise?

Reply:
Ah, I don’t believe so. The StaticModuleRecord passes from the parent compartment to the child compartment strictly through the importHook, which the client code cannot see. With this change, anyone with the SMR can see the code that made it (indirectly) and also infer the module transform (which is not secret). Generally, SMRs only exist in layers where the holder also already has access to the source.

The greatest danger to making this stuff public is not to security but to unintended coupling. We can only advise so much with underbars, but changes to the duck-type will require more care.